File: typescript/src/pfid.ts
import { randomBytes } from 'crypto';
import { PfidError } from './errors';

// Type definitions
export type BinaryPfid = Buffer; // 20 bytes (160 bits)
export type Partition = number; // 0 to 1,073,741,823
export type Timestamp = number; // 0 to 281,474,976,710,655
export type Pfid = string; // 32 character Crockford Base32 string

// Constants
const MAX_TIMESTAMP = 281_474_976_710_655; // 2^48 - 1
const MAX_PARTITION = 1_073_741_823; // 2^30 - 1

// Crockford Base32 encoding/decoding
const ENCODE_CHARS = '0123456789abcdefghjkmnpqrstvwxyz';
const DECODE_MAP: Record<string, number> = {
  '0': 0, '1': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7,
  '8': 8, '9': 9, 'a': 10, 'b': 11, 'c': 12, 'd': 13, 'e': 14, 'f': 15,
  'g': 16, 'h': 17, 'j': 18, 'k': 19, 'm': 20, 'n': 21, 'p': 22, 'q': 23,
  'r': 24, 's': 25, 't': 26, 'v': 27, 'w': 28, 'x': 29, 'y': 30, 'z': 31,
};

function encodeChar(value: number): string {
  return ENCODE_CHARS[value];
}

function decodeChar(char: string): number {
  const value = DECODE_MAP[char.toLowerCase()];
  if (value === undefined) {
    throw new Error(`Invalid character: ${char}`);
  }
  return value;
}

// Validation guards
function isValidTimestamp(timestamp: number): boolean {
  return Number.isInteger(timestamp) && timestamp >= 0 && timestamp <= MAX_TIMESTAMP;
}

function isValidPartition(partition: number): boolean {
  return Number.isInteger(partition) && partition >= 0 && partition <= MAX_PARTITION;
}

/**
 * A zero PFID -- probably don't actually use it, but if you need a placeholder.
 */
export function zero(): Pfid {
  return '00000000000000000000000000000000';
}

/**
 * Generate a Crockford Base32 encoded PFID string with current time.
 */
export function generate(partition: Partition): Pfid {
  if (!isValidPartition(partition)) {
    throw new Error(`Invalid partition: ${partition}`);
  }
  return unsafeEncode(generateBinary(partition));
}

/**
 * Generate a Crockford Base32 encoded PFID string with a provided Unix timestamp.
 */
export function generateWithTimestamp(partition: Partition, timestamp: Timestamp): Pfid {
  if (!isValidPartition(partition)) {
    throw new Error(`Invalid partition: ${partition}`);
  }
  if (!isValidTimestamp(timestamp)) {
    throw new Error(`Invalid timestamp: ${timestamp}`);
  }
  return unsafeEncode(generateBinaryWithTimestamp(partition, timestamp));
}

/**
 * Generate an ID suitable for use in an example -- it's well into the past.
 */
export function generateExample(): Pfid {
  return generateWithTimestamp(123_456_789, 1_234_567_890_000);
}

/**
 * Generate an ID with the same partition as an existing PFID.
 */
export function generateRelated(existingPfid: Pfid): Pfid {
  return generate(extractPartitionOrThrow(existingPfid));
}

/**
 * Generate an ID with a random partition.
 */
export function generateRoot(): Pfid {
  return generate(generatePartition());
}

/**
 * Generate a binary PFID with current time.
 */
export function generateBinary(partition: Partition): BinaryPfid {
  if (!isValidPartition(partition)) {
    throw new Error(`Invalid partition: ${partition}`);
  }
  const timestamp = Date.now();
  return generateBinaryWithTimestamp(partition, timestamp);
}

/**
 * Generate a binary PFID with a provided Unix timestamp.
 * 
 * Binary layout (20 bytes):
 * - Bytes 0-5: timestamp (48 bits, big-endian)
 * - Bytes 6-9: partition (32 bits, big-endian, but only 30 bits used)
 * - Bytes 10-19: randomness (80 bits)
 */
export function generateBinaryWithTimestamp(partition: Partition, timestamp: Timestamp): BinaryPfid {
  if (!isValidPartition(partition)) {
    throw new Error(`Invalid partition: ${partition}`);
  }
  if (!isValidTimestamp(timestamp)) {
    throw new Error(`Invalid timestamp: ${timestamp}`);
  }

  const buffer = Buffer.allocUnsafe(20);

  // Write timestamp (48 bits = 6 bytes) - big endian
  // JavaScript bitwise operators only work on 32 bits, so we need to use division for high bits
  buffer[0] = Math.floor(timestamp / 0x10000000000) & 0xff; // >> 40
  buffer[1] = Math.floor(timestamp / 0x100000000) & 0xff;   // >> 32
  buffer[2] = (timestamp >> 24) & 0xff;
  buffer[3] = (timestamp >> 16) & 0xff;
  buffer[4] = (timestamp >> 8) & 0xff;
  buffer[5] = timestamp & 0xff;

  // Write partition (32 bits = 4 bytes) - big endian
  buffer.writeUInt32BE(partition, 6);

  // Write randomness (80 bits = 10 bytes)
  randomBytes(10).copy(buffer, 10);

  return buffer;
}

/**
 * Check if a string is a valid PFID.
 * Valid PFID: 32 characters, first character must be 0-7, rest must be valid Crockford Base32.
 */
export function isPfid(string: unknown): boolean {
  if (typeof string !== 'string') {
    return false;
  }

  if (string.length !== 32) {
    return false;
  }

  // First character must be 0-7
  if (string[0] < '0' || string[0] > '7') {
    return false;
  }

  // All characters must be valid Crockford Base32
  const validPattern = /^[0-7][0-9abcdefghjkmnpqrstvwxyz]{31}$/;
  return validPattern.test(string);
}

/**
 * Encode a binary PFID to a Crockford Base32 string.
 */
export function encode(binary: BinaryPfid): { ok: true; value: Pfid } | { ok: false; error: PfidError } {
  if (!Buffer.isBuffer(binary) || binary.length !== 20) {
    return { ok: false, error: PfidError.make('invalid_binary', binary) };
  }

  try {
    const encoded = unsafeEncode(binary);
    if (isPfid(encoded)) {
      return { ok: true, value: encoded };
    } else {
      return { ok: false, error: PfidError.make('invalid_binary', binary) };
    }
  } catch (error) {
    return { ok: false, error: PfidError.make('invalid_binary', binary) };
  }
}

/**
 * Decode a Crockford Base32 PFID string to binary.
 */
export function decode(pfid: string): { ok: true; value: BinaryPfid } | { ok: false; error: PfidError } {
  if (typeof pfid !== 'string' || pfid.length !== 32) {
    return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
  }

  // First character must be 0-7
  if (pfid[0] < '0' || pfid[0] > '7') {
    return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
  }

  try {
    const binary = unsafeDecode(pfid);
    return { ok: true, value: binary };
  } catch (error) {
    return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
  }
}

/**
 * Extract partition from a PFID string.
 */
export function extractPartition(pfid: string): { ok: true; value: Partition } | { ok: false; error: PfidError } {
  if (!isPfid(pfid)) {
    return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
  }

  try {
    // Extract the partition portion (characters 10-15, 6 characters = 30 bits)
    const partitionStr = pfid.substring(10, 16);
    const result = decodePartition(partitionStr);
    if (result.ok) {
      return result;
    } else {
      return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
    }
  } catch (error) {
    return { ok: false, error: PfidError.make('invalid_pfid', pfid) };
  }
}

/**
 * Extract partition from a PFID string, throwing on error.
 */
export function extractPartitionOrThrow(pfid: string): Partition {
  const result = extractPartition(pfid);
  if (result.ok) {
    return result.value;
  } else {
    throw result.error;
  }
}

/**
 * Generate a random partition.
 */
export function generatePartition(): Partition {
  const buffer = randomBytes(4);
  // Clear top 2 bits to ensure partition is in valid range
  buffer[0] = buffer[0] & 0x3f; // Keep only bottom 6 bits of first byte
  // Read as 32-bit unsigned integer, but only use 30 bits
  const partition = buffer.readUInt32BE(0) & 0x3fffffff;
  return partition;
}

// Internal encoding/decoding functions

/**
 * Unsafe encode - assumes valid 20-byte buffer.
 * Encodes 160 bits as 32 characters of Crockford Base32.
 * 
 * Bit layout (reading from binary):
 * - t1::3, t2::5, t3::5, t4::5, t5::5, t6::5, t7::5, t8::5, t9::5, t10::5 (48 bits timestamp)
 * - 0::2 (2 bits padding)
 * - p1::5, p2::5, p3::5, p4::5, p5::5, p6::5 (30 bits partition)
 * - r1::5, r2::5, r3::5, r4::5, r5::5, r6::5, r7::5, r8::5, r9::5, r10::5, r11::5, r12::5, r13::5, r14::5, r15::5, r16::5 (80 bits randomness)
 */
function unsafeEncode(binary: Buffer): Pfid {
  // Helper to read bits across byte boundaries
  let bitOffset = 0;

  function readBits(count: number): number {
    let value = 0;
    for (let i = 0; i < count; i++) {
      const byteIndex = Math.floor(bitOffset / 8);
      const bitIndex = 7 - (bitOffset % 8);
      const bit = (binary[byteIndex] >> bitIndex) & 1;
      value = (value << 1) | bit;
      bitOffset++;
    }
    return value;
  }

  // Read timestamp: 3 + 5*9 = 48 bits
  const t1 = readBits(3);
  const t2 = readBits(5);
  const t3 = readBits(5);
  const t4 = readBits(5);
  const t5 = readBits(5);
  const t6 = readBits(5);
  const t7 = readBits(5);
  const t8 = readBits(5);
  const t9 = readBits(5);
  const t10 = readBits(5);

  // Skip 2 padding bits
  readBits(2);

  // Read partition: 5*6 = 30 bits
  const p1 = readBits(5);
  const p2 = readBits(5);
  const p3 = readBits(5);
  const p4 = readBits(5);
  const p5 = readBits(5);
  const p6 = readBits(5);

  // Read randomness: 5*16 = 80 bits
  const r1 = readBits(5);
  const r2 = readBits(5);
  const r3 = readBits(5);
  const r4 = readBits(5);
  const r5 = readBits(5);
  const r6 = readBits(5);
  const r7 = readBits(5);
  const r8 = readBits(5);
  const r9 = readBits(5);
  const r10 = readBits(5);
  const r11 = readBits(5);
  const r12 = readBits(5);
  const r13 = readBits(5);
  const r14 = readBits(5);
  const r15 = readBits(5);
  const r16 = readBits(5);

  return (
    encodeChar(t1) + encodeChar(t2) + encodeChar(t3) + encodeChar(t4) + encodeChar(t5) +
    encodeChar(t6) + encodeChar(t7) + encodeChar(t8) + encodeChar(t9) + encodeChar(t10) +
    encodeChar(p1) + encodeChar(p2) + encodeChar(p3) + encodeChar(p4) + encodeChar(p5) + encodeChar(p6) +
    encodeChar(r1) + encodeChar(r2) + encodeChar(r3) + encodeChar(r4) + encodeChar(r5) + encodeChar(r6) +
    encodeChar(r7) + encodeChar(r8) + encodeChar(r9) + encodeChar(r10) + encodeChar(r11) + encodeChar(r12) +
    encodeChar(r13) + encodeChar(r14) + encodeChar(r15) + encodeChar(r16)
  );
}

/**
 * Unsafe decode - assumes valid 32-character string.
 * Decodes 32 characters of Crockford Base32 to 160 bits (20 bytes).
 * 
 * The decode process reads each character as a byte, decodes it to a 5-bit value,
 * then packs those bits back into the binary format.
 */
function unsafeDecode(pfid: string): BinaryPfid {
  // Decode each character to its 5-bit value (except t1 which is 3 bits)
  const t1 = decodeChar(pfid[0]);  // 3 bits
  const t2 = decodeChar(pfid[1]);   // 5 bits
  const t3 = decodeChar(pfid[2]);   // 5 bits
  const t4 = decodeChar(pfid[3]);   // 5 bits
  const t5 = decodeChar(pfid[4]);   // 5 bits
  const t6 = decodeChar(pfid[5]);   // 5 bits
  const t7 = decodeChar(pfid[6]);   // 5 bits
  const t8 = decodeChar(pfid[7]);   // 5 bits
  const t9 = decodeChar(pfid[8]);   // 5 bits
  const t10 = decodeChar(pfid[9]);  // 5 bits

  const p1 = decodeChar(pfid[10]);  // 5 bits
  const p2 = decodeChar(pfid[11]);  // 5 bits
  const p3 = decodeChar(pfid[12]);  // 5 bits
  const p4 = decodeChar(pfid[13]);  // 5 bits
  const p5 = decodeChar(pfid[14]);  // 5 bits
  const p6 = decodeChar(pfid[15]);  // 5 bits

  const r1 = decodeChar(pfid[16]);  // 5 bits
  const r2 = decodeChar(pfid[17]);  // 5 bits
  const r3 = decodeChar(pfid[18]);  // 5 bits
  const r4 = decodeChar(pfid[19]);  // 5 bits
  const r5 = decodeChar(pfid[20]);  // 5 bits
  const r6 = decodeChar(pfid[21]);  // 5 bits
  const r7 = decodeChar(pfid[22]);  // 5 bits
  const r8 = decodeChar(pfid[23]);  // 5 bits
  const r9 = decodeChar(pfid[24]);  // 5 bits
  const r10 = decodeChar(pfid[25]); // 5 bits
  const r11 = decodeChar(pfid[26]); // 5 bits
  const r12 = decodeChar(pfid[27]); // 5 bits
  const r13 = decodeChar(pfid[28]); // 5 bits
  const r14 = decodeChar(pfid[29]); // 5 bits
  const r15 = decodeChar(pfid[30]); // 5 bits
  const r16 = decodeChar(pfid[31]); // 5 bits

  // Helper to write bits across byte boundaries
  const buffer = Buffer.alloc(20, 0);
  let bitOffset = 0;

  function writeBits(value: number, count: number): void {
    for (let i = count - 1; i >= 0; i--) {
      const byteIndex = Math.floor(bitOffset / 8);
      const bitIndex = 7 - (bitOffset % 8);
      const bit = (value >> i) & 1;
      buffer[byteIndex] |= (bit << bitIndex);
      bitOffset++;
    }
  }

  // Write timestamp: 3 + 5*9 = 48 bits
  writeBits(t1, 3);
  writeBits(t2, 5);
  writeBits(t3, 5);
  writeBits(t4, 5);
  writeBits(t5, 5);
  writeBits(t6, 5);
  writeBits(t7, 5);
  writeBits(t8, 5);
  writeBits(t9, 5);
  writeBits(t10, 5);

  // Write 2 padding bits (0)
  writeBits(0, 2);

  // Write partition: 5*6 = 30 bits
  writeBits(p1, 5);
  writeBits(p2, 5);
  writeBits(p3, 5);
  writeBits(p4, 5);
  writeBits(p5, 5);
  writeBits(p6, 5);

  // Write randomness: 5*16 = 80 bits
  writeBits(r1, 5);
  writeBits(r2, 5);
  writeBits(r3, 5);
  writeBits(r4, 5);
  writeBits(r5, 5);
  writeBits(r6, 5);
  writeBits(r7, 5);
  writeBits(r8, 5);
  writeBits(r9, 5);
  writeBits(r10, 5);
  writeBits(r11, 5);
  writeBits(r12, 5);
  writeBits(r13, 5);
  writeBits(r14, 5);
  writeBits(r15, 5);
  writeBits(r16, 5);

  return buffer;
}

/**
 * Decode partition from 6-character encoded partition string.
 */
function decodePartition(partitionStr: string): { ok: true; value: Partition } | { ok: false; error: PfidError } {
  if (typeof partitionStr !== 'string' || partitionStr.length !== 6) {
    return { ok: false, error: PfidError.make('invalid_partition', partitionStr) };
  }

  try {
    const p1 = decodeChar(partitionStr[0]);
    const p2 = decodeChar(partitionStr[1]);
    const p3 = decodeChar(partitionStr[2]);
    const p4 = decodeChar(partitionStr[3]);
    const p5 = decodeChar(partitionStr[4]);
    const p6 = decodeChar(partitionStr[5]);

    // Reconstruct the 30-bit partition value
    // The partition is stored as: 0 (2 bits) + p1 (5 bits) + p2 (5 bits) + ... + p6 (5 bits)
    const buffer = Buffer.allocUnsafe(4);
    buffer[0] = (0 << 6) | (p1 << 1) | (p2 >> 4);
    buffer[1] = ((p2 & 0x0f) << 4) | (p3 >> 1);
    buffer[2] = ((p3 & 0x01) << 7) | (p4 << 2) | (p5 >> 3);
    buffer[3] = ((p5 & 0x07) << 5) | (p6 >> 0);

    // Read as 32-bit unsigned integer, mask to 30 bits
    const partition = buffer.readUInt32BE(0) & 0x3fffffff;
    return { ok: true, value: partition };
  } catch (error) {
    return { ok: false, error: PfidError.make('invalid_partition', partitionStr) };
  }
}

File: elixir/lib/pfid.ex
# A fork of the code in https://github.com/vonagam/ash_ulid which seems to be the best implementation but also
# depends on the whole framework.  Modified heavily, it's not even a ulid anymore.

# ORIGINAL LICENSE FOR THIS FILE:

# MIT License
# Copyright (c) 2023 Dmitry Maganov
# Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation
# files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy,
# modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software
# is furnished to do so, subject to the following conditions:
# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
# OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
# IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# MODIFICATIONS BEYOND THE ORIGINAL FILE Copyright (c) 2025, Prefactor Pty Ltd

# Our format is like this:

# 01an4z07byd9df0k79ka1307sr9x4mv3
# |--------||----||--------------|
#  Time      Part  Randomness
#  48bits    30b   80bits

# The "part" is "partition" and is 30 bits.
# The other sections are just like normal ulid

# Bit length = 160 (48 + 32 + 80)  Note that the partition field is 32 bits, but only 30 bits are used.  We're maintaining byte alignment for the fields.
# Character length = 32 (256 bits)

defmodule Pfid do
  defmodule PfidError do
    defexception [:code, :message]

    def make(code, problem: problem) do
      %PfidError{
        code: code,
        message:
          case code do
            :invalid_binary -> "invalid binary PFID: #{inspect(problem)}"
            :invalid_pfid -> "invalid PFID: #{inspect(problem)}"
            :invalid_partition -> "invalid partition: #{inspect(problem)}"
          end
      }
    end
  end

  defguard valid_timestamp?(timestamp)
           when is_integer(timestamp) and timestamp >= 0 and timestamp < 281_474_976_710_656

  defguard valid_partition?(partition)
           when is_integer(partition) and partition >= 0 and partition < 1_073_741_824

  @moduledoc """
  Helpers for working with PFIDs.
  """

  @typedoc """
  A Crockford Base32 encoded PFID string.
  """
  @type t :: <<_::256>>

  @typedoc """
  A raw binary representation of a PFID.
  """
  @type binary_pfid :: <<_::160>>

  @typedoc """
  A partition number for a PFID.
  """
  @type partition :: 0..1_073_741_823

  @typedoc """
  A timestamp for a PFID.
  """
  @type timestamp :: 0..281_474_976_710_655

  @doc """
  A zero PFID -- probably don't actually use it, but if you need a placeholder.
  """
  @spec zero() :: t
  def zero() do
    "00000000000000000000000000000000"
  end

  @doc """
  Generate a Crockford Base32 encoded PFID string with current time.
  """
  @spec generate(partition) :: t
  def generate(partition) when valid_partition?(partition) do
    unsafe_encode(generate_binary(partition))
  end

  @doc """
  Generate a Crockford Base32 encoded PFID string with a provided Unix timestamp.
  """
  @spec generate(partition, timestamp) :: t
  def generate(partition, timestamp)
      when valid_partition?(partition) and valid_timestamp?(timestamp) do
    unsafe_encode(generate_binary(partition, timestamp))
  end

  @doc """
  Generate an ID suitable for use in an example -- it's well into the past.
  """
  def generate_example() do
    generate(123_456_789, 1_234_567_890_000)
  end

  @doc """
  Generate an ID with the same partition as an existing PFID.
  """
  def generate_related(existing_pfid) do
    generate(extract_partition!(existing_pfid))
  end

  @doc """
  Generate an ID with a random partition.
  """
  def generate_root() do
    generate(generate_partition())
  end

  @doc """
  Generate a binary PFID with current time.
  """
  @spec generate_binary(partition) :: binary_pfid
  def generate_binary(partition) when valid_partition?(partition) do
    <<System.system_time(:millisecond)::unsigned-size(48), partition::unsigned-size(32),
      :crypto.strong_rand_bytes(10)::binary>>
  end

  @doc """
  Generate a binary PFID with a provided Unix timestamp.
  """
  @spec generate_binary(partition, timestamp) :: binary_pfid
  def generate_binary(partition, timestamp)
      when valid_partition?(partition) and valid_timestamp?(timestamp) do
    <<timestamp::unsigned-size(48), partition::unsigned-size(32),
      :crypto.strong_rand_bytes(10)::binary>>
  end

  @spec is_pfid?(string :: any) :: boolean()
  def is_pfid?(string) when is_binary(string) do
    String.match?(string, ~r/^[0-7][0-9abcdefghjkmnpqrstvwxyz]{31}$/)
  end

  def is_pfid?(_) do
    false
  end

  defp unsafe_encode(
         <<t1::3, t2::5, t3::5, t4::5, t5::5, t6::5, t7::5, t8::5, t9::5, t10::5, 0::2, p1::5,
           p2::5, p3::5, p4::5, p5::5, p6::5, r1::5, r2::5, r3::5, r4::5, r5::5, r6::5, r7::5,
           r8::5, r9::5, r10::5, r11::5, r12::5, r13::5, r14::5, r15::5, r16::5>>
       ) do
    <<e(t1), e(t2), e(t3), e(t4), e(t5), e(t6), e(t7), e(t8), e(t9), e(t10), e(p1), e(p2), e(p3),
      e(p4), e(p5), e(p6), e(r1), e(r2), e(r3), e(r4), e(r5), e(r6), e(r7), e(r8), e(r9), e(r10),
      e(r11), e(r12), e(r13), e(r14), e(r15), e(r16)>>
  end

  def encode(binary) when is_binary(binary) and byte_size(binary) == 20 do
    encoded = unsafe_encode(binary)

    if is_pfid?(encoded) do
      {:ok, encoded}
    else
      {:error, PfidError.make(:invalid_binary, problem: binary)}
    end
  end

  def encode(value) do
    {:error, PfidError.make(:invalid_binary, problem: value)}
  end

  def decode(
        <<t1::8, t2::8, t3::8, t4::8, t5::8, t6::8, t7::8, t8::8, t9::8, t10::8, p1::8, p2::8,
          p3::8, p4::8, p5::8, p6::8, r1::8, r2::8, r3::8, r4::8, r5::8, r6::8, r7::8, r8::8,
          r9::8, r10::8, r11::8, r12::8, r13::8, r14::8, r15::8, r16::8>> = binary
      )
      when t1 < ?8 do
    {:ok,
     <<d(t1)::3, d(t2)::5, d(t3)::5, d(t4)::5, d(t5)::5, d(t6)::5, d(t7)::5, d(t8)::5, d(t9)::5,
       d(t10)::5, 0::2, d(p1)::5, d(p2)::5, d(p3)::5, d(p4)::5, d(p5)::5, d(p6)::5, d(r1)::5,
       d(r2)::5, d(r3)::5, d(r4)::5, d(r5)::5, d(r6)::5, d(r7)::5, d(r8)::5, d(r9)::5, d(r10)::5,
       d(r11)::5, d(r12)::5, d(r13)::5, d(r14)::5, d(r15)::5, d(r16)::5>>}
  catch
    :error ->
      {:error, PfidError.make(:invalid_pfid, problem: binary)}
  end

  def decode(value) do
    {:error, PfidError.make(:invalid_pfid, problem: value)}
  end

  def decode_partition(<<p1::8, p2::8, p3::8, p4::8, p5::8, p6::8>> = binary) do
    {:ok,
     :binary.decode_unsigned(<<0::2, d(p1)::5, d(p2)::5, d(p3)::5, d(p4)::5, d(p5)::5, d(p6)::5>>)}
  catch
    :error ->
      {:error, PfidError.make(:invalid_partition, problem: binary)}
  end

  def decode_partition(value) do
    {:error, PfidError.make(:invalid_partition, problem: value)}
  end

  def extract_partition(<<_::80, encoded_partition::binary-size(6), _::128>> = string) do
    if is_pfid?(string) do
      decode_partition(encoded_partition)
    else
      {:error, PfidError.make(:invalid_pfid, problem: string)}
    end
  end

  def extract_partition(value) do
    {:error, PfidError.make(:invalid_pfid, problem: value)}
  end

  def extract_partition!(pfid) do
    case extract_partition(pfid) do
      {:ok, value} -> value
      {:error, error} -> raise error
    end
  end

  def generate_partition() do
    <<_::2, partition::30>> = :crypto.strong_rand_bytes(4)
    partition
  end

  @compile {:inline, e: 1}
  defp e(0), do: ?0
  defp e(1), do: ?1
  defp e(2), do: ?2
  defp e(3), do: ?3
  defp e(4), do: ?4
  defp e(5), do: ?5
  defp e(6), do: ?6
  defp e(7), do: ?7
  defp e(8), do: ?8
  defp e(9), do: ?9
  defp e(10), do: ?a
  defp e(11), do: ?b
  defp e(12), do: ?c
  defp e(13), do: ?d
  defp e(14), do: ?e
  defp e(15), do: ?f
  defp e(16), do: ?g
  defp e(17), do: ?h
  defp e(18), do: ?j
  defp e(19), do: ?k
  defp e(20), do: ?m
  defp e(21), do: ?n
  defp e(22), do: ?p
  defp e(23), do: ?q
  defp e(24), do: ?r
  defp e(25), do: ?s
  defp e(26), do: ?t
  defp e(27), do: ?v
  defp e(28), do: ?w
  defp e(29), do: ?x
  defp e(30), do: ?y
  defp e(31), do: ?z

  @compile {:inline, d: 1}
  defp d(?0), do: 0
  defp d(?1), do: 1
  defp d(?2), do: 2
  defp d(?3), do: 3
  defp d(?4), do: 4
  defp d(?5), do: 5
  defp d(?6), do: 6
  defp d(?7), do: 7
  defp d(?8), do: 8
  defp d(?9), do: 9
  defp d(?a), do: 10
  defp d(?b), do: 11
  defp d(?c), do: 12
  defp d(?d), do: 13
  defp d(?e), do: 14
  defp d(?f), do: 15
  defp d(?g), do: 16
  defp d(?h), do: 17
  defp d(?j), do: 18
  defp d(?k), do: 19
  defp d(?m), do: 20
  defp d(?n), do: 21
  defp d(?p), do: 22
  defp d(?q), do: 23
  defp d(?r), do: 24
  defp d(?s), do: 25
  defp d(?t), do: 26
  defp d(?v), do: 27
  defp d(?w), do: 28
  defp d(?x), do: 29
  defp d(?y), do: 30
  defp d(?z), do: 31
  defp d(_), do: throw(:error)
end
